### ISSUE TITLE
Considerations for GraphQL support

### ISSUE THREAD

**Issue by brylie** [NREL/api-umbrella#297]
@GUI what are the considerations for GraphQL support in API Umbrella? Is it currently possible to make GraphQL requests through API Umbrella?

**Comment by GUI** [comment]
Yup, assuming you're talking [GraphQL over HTTP](http://graphql.org/learn/serving-over-http/) (which is the most common thing at this point), I don't think there's any reason it wouldn't work out of the box (their docs on HTTP even talk about GraphQL being placed behind other middleware layers). GraphQL should behave just like any other HTTP request, with the primary difference being that all the query data is passed in the HTTP body (rather than as query string parameters). So while I haven't tested it, I believe API Umbrella should be fine proxying to any GraphQL backend.

**Comment by GUI** [comment]
A couple more thoughts: There are probably two main differences when proxying to GraphQL backends, though:

1. The HTTP caching layer API Umbrella provides wouldn't be much use, since GraphQL isn't really cacheable by a traditional HTTP caching layer (since all the URLs are the same and there can be a lot of variation in the body of the request). But caching could still be done by the GraphQL server or clients that are built specifically for GraphQL.
2. API Umbrella's analytics will capture all GraphQL hits as requests to the root URL, but you may not have as much insight as to what specific queries are being performed. This is because GraphQL passes all the query data in the HTTP body, which we don't capture or parse. So you'll still have high-level metrics for the GraphQL API, but you may not be able to discern as much about the specific types of queries being performed.

**Comment by tuukka** [comment]
I would add to the list:

3. The cost of a single request isn't bounded. That's why GraphQL APIs would benefit from a cost being returned from the API to API Umbrella in an additional HTTP header for quota purposes.

As to solving analytics (point 2), would it be possible to include a GraphQL parser that could extract the top-level and sub-queries contained in a request?

**Comment by brylie** [comment]
Related issue #308

**Comment by Nosferican** [comment]
What's the status in terms of API powered by Umbrella to implement GraphQL?
